bouncer/shared/fill_or_kill.ts

```diff
@@ -0,0 +1,139 @@
+
+  const observeSwapScheduled = observeEvent('swapping:SwapScheduled', {
+    test: (event) =>
+      typeof event.data.origin === 'object' &&
+      'DepositChannel' in event.data.origin &&
+      event.data.origin.DepositChannel.channelId === depositChannelId.toString() &&
+      event.data.destinationAddress[shortChainFromAsset(swapAsset)]?.toLowerCase() ===
+        observeDestinationAddress.toLowerCase(),
+  }).event;
+
+  // Deposit the asset
+  await send(asset, depositAddress, amount.toString());
+  console.log(`Sent ${amount} ${asset} to ${depositAddress}`);
+
+  const swapScheduledEvent = await observeSwapScheduled;
+  const swapId = Number(swapScheduledEvent.data.swapId.replaceAll(',', ''));
+  console.log(`${asset} swap scheduled, swapId: ${swapId}`);
+
+  // TODO: Observing after the SwapScheduled event means its possible to miss the events, but we need to the swap id.
+  const observeRefundEgressScheduled = observeEvent(`swapping:RefundEgressScheduled`, {
+    test: (event) => Number(event.data.swapId.replaceAll(',', '')) === swapId,
+  }).event;
+  const observeSwapExecuted = observeEvent(`swapping:SwapExecuted`, {
+    test: (event) => Number(event.data.swapId.replaceAll(',', '')) === swapId,
+  }).event;
+  const observeSwapRescheduled = observeEvent(`swapping:SwapRescheduled`, {
+    test: (event) => Number(event.data.swapId.replaceAll(',', '')) === swapId,
+  }).event;
+
+  const firstSwapEvent = await Promise.race([
+    observeSwapRescheduled,
+    observeSwapExecuted,
+    observeRefundEgressScheduled,
+  ]);
+  if (firstSwapEvent.name.method === 'SwapExecuted') {
+    throw new Error(`${asset} swap ${swapId} was executed instead of failing and being retried`);
+  } else if (firstSwapEvent.name.method !== 'SwapRescheduled') {
+    throw new Error(`${asset} swap ${swapId} was not retried once as expected`);
+  }
+  console.log(`${asset} swap ${swapId} was rescheduled`);
+
+  const secondSwapEvent = await Promise.race([observeRefundEgressScheduled, observeSwapExecuted]);
+  if (secondSwapEvent.name.method === 'SwapExecuted') {
+    throw new Error(`${asset} swap ${swapId} was executed instead of being refunded`);
+  }
+
+  const refundBalanceAfter = await observeBalanceIncrease(
+    asset,
+    refundAddress,
+    refundBalanceBefore,
+  );
+
+  console.log(
+    `Refund balance before: ${refundBalanceBefore}, after: ${refundBalanceAfter} ${asset} `,
+  );
+  // We expect the refund to be a little less due to ingress and egress fees.
+  assert(
+    isWithinOnePercent(
+      amountToFineAmountBigInt(refundBalanceAfter, asset),
+      amountToFineAmountBigInt(refundBalanceBefore, asset) +
+        amountToFineAmountBigInt(amount, asset),
+    ),
+    `${asset} refund amount is incorrect (swapId: ${swapId})`,
+  );
+}
+
+export async function testFillOrKill() {
+  console.log('\x1b[36m%s\x1b[0m', '=== Running FoK test ===');
+
+  await Promise.all([
+    testMinPriceRefund(Assets.Flip, 500),
```

Comment:
I know this is going to be controversial, but I don't think we need to test this for multiple assets. I know it is easy to do, but I really think we should be more deliberate about what we are testing. What's important is that FoK works end-to-end. The rest is just a distraction (not to mention other, less important issues like scalability, maintenance etc). Testing multiple assets suggests that there are different code paths for FoK depending on the asset, but that's not the case. If someone adds a different code path for a some asset (which is unlikely), it will be their responsibility to add a test for that (and it will probably be a unit test anyway, not a bouncer test).